packages/api-codec/src/StorageKey.ts

```diff
@@ -2,8 +2,45 @@
 // This software may be modified and distributed under the terms
 // of the ISC license. See the LICENSE file for details.
 
+import { AnyU8a } from './types';
+
+import { StorageFunctionMetadata, StorageFunctionType } from './Metadata';
 import Bytes from './Bytes';
 
-// A representation of a storage key (typically hashed) in the system
+export interface StorageFunction extends StorageFunctionMetadata {
+  create (arg?: any): Uint8Array;
+}
+
+type StorageFunctionWithArg = [StorageFunction, any];
+
+const DEFAULT_TYPE = new StorageFunctionType(0, 'Bytes');
+
+// A representation of a storage key (typically hashed) in the system. It can be constructed
+// by passing in a raw key or a StorageFunction with (optional) arguments.
 export default class StorageKey extends Bytes {
+  private _outputType: StorageFunctionType;
```

Comment:
Not 100% sure on this atm - basically the StorageKey carries around the resulting type. Then when a result comes back, the formatter can jump in and grab it.